File: app/src/main/java/com/donald/musictheoryapp/PanelOnTouchListener.java
package com.donald.musictheoryapp;

import android.view.MotionEvent;
import android.view.View;
import android.widget.TextView;

import com.donald.musictheoryapp.Music.MusicXML.Clef;
import com.donald.musictheoryapp.Music.MusicXML.Measure;
import com.donald.musictheoryapp.Music.MusicXML.Note;
import com.donald.musictheoryapp.Music.MusicXML.Score;
import com.donald.musictheoryapp.Music.ScoreView.ScoreView;
import com.donald.musictheoryapp.Question.IntervalInputQuestion;

import java.util.Arrays;

public class PanelOnTouchListener implements View.OnTouchListener
{
	private final Note inputNote;
	private final ScoreView scoreView;
	private float initialY;
	private int initialAbsStep;
	private final int firstNoteAbsStep;
	private final int firstNoteAlter;
	private final int[] measureAlters;
	private final Clef clef;
	private static final int NOTE_POS_UPPER_LIMIT = 15;
	private static final int NOTE_POS_LOWER_LIMIT = -7;

	public PanelOnTouchListener(ScoreView scoreView, IntervalInputQuestion question)
	{
		Score score = question.score();
		this.inputNote = score.parts()[0].measures()[0].notes()[1];
		this.scoreView = scoreView;
		question.answer.userAnswer = inputNote;
		Measure measure = score.parts()[0].measures()[0];
		firstNoteAbsStep = measure.notes()[0].pitch().absStep();
		//firstNoteAbsStep = measure.notes()[0].pitch().alter();
		firstNoteAlter = measure.notes()[0].pitch().alter();
		clef = measure.attributes.clefs[0];

		measureAlters = new int[Note.Pitch.Step.NO_OF_STEPS];
		Arrays.fill(measureAlters, 0);

		if(measure.attributes.key.fifths != 0)
		{
			int i = 0;
			int currentStep;
			int alter;
			int stepItDisplacement;
			if(measure.attributes.key.fifths > 0)
			{
				// sharping from F
				currentStep = Note.Pitch.Step.F;
				alter = 1;
				stepItDisplacement = 4;
			}
			else
			{
				// flatting from B
				currentStep = Note.Pitch.Step.B;
				alter = -1;
				stepItDisplacement = -3;
			}
			int fifths = Math.abs(measure.attributes.key.fifths);
			while(i < fifths)
			{
				measureAlters[currentStep] += alter;
				i++;
				currentStep = Math.floorMod(currentStep + stepItDisplacement, Note.Pitch.Step.NO_OF_STEPS);
			}
		}
	}

	@Override
	public boolean onTouch(View view, MotionEvent event)
	{
		int action = event.getActionMasked();
		switch(action)
		{
			// TODO: WHAT IF ACTION_MOVE HAPPENS BEFORE ACTION_DOWN
			case MotionEvent.ACTION_DOWN:
			{
				initialY = event.getY();
				initialAbsStep = inputNote.pitch().absStep();
				return true;
			}
			case MotionEvent.ACTION_UP:
			{
				float currentY = event.getY();
				float distance = currentY - initialY;
				if(event.getEventTime() - event.getDownTime() < 200d && Math.abs((int) distance / 60) < 1)
				{
					int accidental =
						((inputNote.accidental() + 1 + 1) % (Note.Accidental.NO_OF_ACCIDENTALS + 1)) - 1;
					inputNote.setAccidental(accidental);
					if(accidental == Note.Accidental.NULL)
					{
						if(inputNote.pitch().absStep() == firstNoteAbsStep)
						{
							inputNote.pitch().setAlter(firstNoteAlter);
						}
						else
						{
							inputNote.pitch().setAlter(measureAlters[inputNote.pitch().step()]);
						}
					}
					else
					{
						inputNote.pitch().setAlter(Note.Accidental.alter(accidental));
					}
					scoreView.postInvalidate();
				}
				return true;
			}
			case MotionEvent.ACTION_MOVE:
			{
				float currentY = event.getY();
				float distance = currentY - initialY;
				int absStepUpperLimit = Clef.Sign.BASE_NOTE_POS_BY_SIGN[clef.sign] + NOTE_POS_UPPER_LIMIT - (clef.line - 1) * 2;
				int absStepLowerLimit = Clef.Sign.BASE_NOTE_POS_BY_SIGN[clef.sign] + NOTE_POS_LOWER_LIMIT - (clef.line - 1) * 2;
				int newAbsStep =
					Math.min(
						Math.max(
							initialAbsStep - (int) distance / 60,
							absStepLowerLimit
						),
						absStepUpperLimit
					);
				inputNote.pitch().setStep(Math.floorMod(newAbsStep, Note.Pitch.Step.NO_OF_STEPS));
				inputNote.pitch().setOctave(newAbsStep / Note.Pitch.Step.NO_OF_STEPS);
				scoreView.postInvalidate();
				return true;
			}
			default:
			{
				return view.performClick();
			}
		}
	}
}

File: app/src/main/java/com/donald/musictheoryapp/listener/PanelOnTouchListener.kt
package com.donald.musictheoryapp.listener

import android.util.Log
import com.donald.musictheoryapp.customview.scoreview.ScoreView
import android.view.View.OnTouchListener
import android.view.MotionEvent
import android.view.View
import com.donald.musictheoryapp.music.musicxml.*
import com.donald.musictheoryapp.music.musicxml.Note
import com.google.android.material.math.MathUtils.floorMod
import java.util.*
import kotlin.math.abs

class PanelOnTouchListener(scoreView: ScoreView, score: Score) : OnTouchListener {

    private val inputNote: Note
    private val scoreView: ScoreView
    private var initialY = 0f
    private var initialAbsStep = 0
    private val firstNoteAbsStep: Int
    private val firstNoteAlter: Int
    private val measureAlters: IntArray
    private val clef: Clef

    override fun onTouch(view: View, event: MotionEvent): Boolean {
        return when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                initialY = event.y
                initialAbsStep = inputNote.pitch!!.absStep
                true
            }
            MotionEvent.ACTION_UP -> {
                Log.d("inputNote alter before", inputNote.pitch!!.alter.toString())
                val currentY = event.y
                val distance = currentY - initialY
                if (event.eventTime - event.downTime < 200.0 && abs(distance.toInt() / 60) < 1) {
                    // next accidental
                    val accidental = inputNote.accidental.next()
                    inputNote.accidental = accidental
                    if (accidental == null) {
                        if (inputNote.pitch.absStep == firstNoteAbsStep) {
                            inputNote.pitch.alter = firstNoteAlter
                        } else {
                            inputNote.pitch.alter = measureAlters[inputNote.pitch.step.ordinal]
                        }
                    } else {
                        inputNote.pitch.alter = accidental.alter
                    }
                    scoreView.invalidate()
                    //scoreView.requestLayout();
                    Log.d("inputNote alter after", inputNote.pitch.alter.toString())
                }
                true
            }
            MotionEvent.ACTION_MOVE -> {
                val currentY = event.y
                val distance = currentY - initialY
                val absStepUpperLimit = clef.sign.baseAbsStep + NOTE_POS_UPPER_LIMIT - (clef.line - 1) * 2
                val absStepLowerLimit = clef.sign.baseAbsStep + NOTE_POS_LOWER_LIMIT - (clef.line - 1) * 2
                val newAbsStep = (initialAbsStep - distance.toInt() / 60).coerceAtLeast(absStepLowerLimit).coerceAtMost(absStepUpperLimit)
                inputNote.pitch?.step = Step.values()[floorMod(newAbsStep, Step.values().size)]
                inputNote.pitch?.octave = newAbsStep / Step.values().size
                scoreView.invalidate()
                true
            }
            else -> {
                view.performClick()
            }
        }
    }

    init {
        inputNote = score.parts()[0].measures()[0].notes()[1]
        this.scoreView = scoreView
        //question.answer.userAnswer = inputNote;
        val measure = score.parts()[0].measures()[0]
        firstNoteAbsStep = measure.notes()[0].pitch!!.absStep
        firstNoteAlter = measure.notes()[0].pitch!!.alter
        clef = measure.attributes!!.clefs[0]
        measureAlters = IntArray(Step.values().size)
        Arrays.fill(measureAlters, 0)
        if (measure.attributes.key.fifths != 0) {
            var i = 0
            var currentStep: Step
            val alter: Int
            val stepItDisplacement: Int
            if (measure.attributes.key.fifths > 0) {
                // sharping from F
                currentStep = Step.F
                alter = 1
                stepItDisplacement = 4
            } else {
                // flatting from B
                currentStep = Step.B
                alter = -1
                stepItDisplacement = -3
            }
            val fifths = abs(measure.attributes.key.fifths)
            while (i < fifths) {
                measureAlters[currentStep.ordinal] += alter
                i++
                currentStep = Step.values()[floorMod(currentStep.ordinal + stepItDisplacement, Step.values().size)]
            }
        }
    }

    companion object {

        private const val NOTE_POS_UPPER_LIMIT = 15
        private const val NOTE_POS_LOWER_LIMIT = -7

        val accidentalCycle = arrayOf(
            null,
            Accidental.FlatFlat,
            Accidental.Flat,
            Accidental.Natural,
            Accidental.Sharp,
            Accidental.SharpSharp
        )

    }

    fun Accidental?.next(step: Int = 1): Accidental? {
        val state = this?.let { ordinal + 1 } ?: 0
        return accidentalCycle[floorMod((state + step), accidentalCycle.size)]
    }

    fun Accidental?.prev(step: Int = 1): Accidental? {
        val state = this?.let { ordinal + 1 } ?: 0
        return accidentalCycle[floorMod((state - step), accidentalCycle.size)]
    }

}